### DISCUSSION TITLE
Assistance to debug plugin

### DISCUSSION THREAD

**Discussion by mpet** [eclipse-m2e/m2e-core#654] [Q&A] [answered]
Hi,
I managed to import the m2e plugin into Eclipse 4.22.
I only see a few errors but in test projects.

![image](https://user-images.githubusercontent.com/1445826/164417984-4839a6fb-cc0e-48c0-8840-24f5b4bb2edc.png)

I see the problem in META-INF files for these maven modules.

Module                                                            Error package
=====                                                           =========    
/org.eclipse.m2e.editor.lemminx.tests               org.eclipse.ui.tests.harness
/org.eclipse.m2e.editor.tests                            org.eclipse.ui.tests.harness;bundle-version="1.8.0"
/org.eclipse.m2e.jdt.tests                                 org.eclipse.ui.tests.harness;bundle-version="1.8.0"

So I guess it is nothing serious for running the plugin.

How can I start Eclipse with this development version installed in debug mode?

br,

//mike

**Comment by laeubi** [comment, marked as answer]
> So I guess it is nothing serious for running the plugin.

You can close these bundles if you only like to debug/develop things without running the full test-suite.

> How can I start Eclipse with this development version installed in debug mode?

Simply create a new Run Configuration of Type "Eclipse Application"

**Reply by HannesWell** [reply]
> > How can I start Eclipse with this development version installed in debug mode?
> 
> Simply create a new Run Configuration of Type "Eclipse Application"

Such a run configuration is already pre-configured when you have opened the m2e-core project in your workspace.
Its name is `Eclipse-M2E-IDE`. You can find it when selecting in the drop-down next to the Debug-Button `Debug Configurations...` then it should be listed on the left in the group `Eclipse Application`. If you have launched it recently you can select it directly in the drop-down menu of the of the Debug-Button.

> I managed to import the m2e plugin into Eclipse 4.22.

If possible we recommend to use the latest release version of Eclipse or even better the latest version (which includes Milestones).
Especially the mentioned launch configuration will only work well with Eclipse 2022-03/4.23.

**Reply by mpet** [reply]
@laeubi  I manged to get debugger working. Thanks!

Since I have annotation processor in one maven module and processor-user in another maven module the annotation processor must be compiled before using it in the processor-user. I checked the .factorypath for annotation processor but I cannot find it. I have done a Maven --> Update from within Eclipse but the dependent module is not added to .factorypath as I expected. How is it decided what goes into the .factorypath?

@HannesWell I tried yours as well and it worked fine in debug mode. 
I also tried 4.23 first but I got compiler errors in the project ( not tests this time) and it was too big threshold to get over for a first timer of m2e plugin :smiley:

**Reply by mpet** [reply]
What I don't understand when I have the annotation-processor as a maven module it is not picked up in the processor user module to generate classes. So there is not clear what I need to define in Eclipse (m2e-apt part) and what needs to be defined in pom.xml for the maven-compiler-plugin

**Comment by mpet** [comment]
@fbricon I turn to you since you have done a lot of commit in AbstractAptConfiguratorDelegate and MavenCompilerJdtAptDelegate.
When I do Maven-->Update and MavenCompilerJdtAptDelegate.getAnnotationProcessorConfiguration(IProgressMonitor monitor) is called I can see that:
```
  final Dependency[] annotationProcessorPaths = maven.getMojoParameterValue(mavenProject, mojoExecution,
          "annotationProcessorPaths", Dependency[].class, monitor);
```
is an empty array.

I have added the following in my maven module using the annotation processor:
``` 
 <dependencies>
        <dependency>
            <groupId>com.baeldung</groupId>
            <artifactId>annotation-processing</artifactId>
            <version>${project.parent.version}</version>
        </dependency>
    </dependencies>

<build>
    <plugins>
        <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-compiler-plugin</artifactId>
            <version>${maven-compiler-plugin.version}</version>
            <configuration>
                <source>1.8</source>
                <target>1.8</target>
                <encoding>UTF-8</encoding>
                <annotationProcessors>
                    <annotationProcessor>
                       com.baeldung.annotation.processor.BuilderProcessor
                    </annotationProcessor>
                </annotationProcessors>
            </configuration>
         </plugin>       	
    </plugins>
</build>
```
Any ideas why it is not picked up?

**Reply by mpet** [reply]
I realized my mistake 😅 It should be like this:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <artifactId>annotation-user</artifactId>
    <name>annotation-user</name>

    <parent>
        <groupId>com.baeldung</groupId>
        <artifactId>annotations</artifactId>
        <version>1.0.0-SNAPSHOT</version>
    </parent>
    
  

    <dependencies>
        <dependency>
            <groupId>com.baeldung</groupId>
            <artifactId>annotation-processing</artifactId>
            <version>${project.parent.version}</version>
        </dependency>
    </dependencies>
    
     <build>
    <plugins>
        <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-compiler-plugin</artifactId>
            <version>${maven-compiler-plugin.version}</version>
            <configuration>
                <source>1.8</source>
                <target>1.8</target>
                <encoding>UTF-8</encoding>
            <annotationProcessorPaths>
   	 		<path>
      		<groupId>com.baeldung</groupId>
      		<artifactId>annotation-processing</artifactId>
      		<version>${project.version}</version>
    		</path>
  			</annotationProcessorPaths>
  			</configuration>
         </plugin>       	
    </plugins>
</build>
```
Now I can see that the annotationProcessorPaths is picked up !

**Reply by mpet** [reply]
I debugged the class, MavenCompilerJdtAptDelegate.getAnnotationProcessorConfiguration(IProgressMonitor monitor)

I can see the content of:
 configuration.setDependencies(dependencies);

[C:\Users\eraonel\git\tutorials\annotations\annotation-processing\target\classes,
 C:\repo\eraonel\.m2\repository\org\slf4j\slf4j-api\1.7.32\slf4j-api-1.7.32.jar,
 C:\repo\eraonel\.m2\repository\ch\qos\logback\logback-classic\1.2.6\logback-classic-1.2.6.jar,
 C:\repo\eraonel\.m2\repository\ch\qos\logback\logback-core\1.2.6\logback-core-1.2.6.jar,
 C:\repo\eraonel\.m2\repository\org\slf4j\jcl-over-slf4j\1.7.32\jcl-over-slf4j-1.7.32.jar]

So my annotation processing classes are included.

Then in AbstractAptConfiguratorDelegate.configureProject(IProgressMonitor monitor) the configuration is referenced.

AnnotationProcessorConfiguration configuration = getAnnotationProcessorConfiguration(monitor);

In the same class jar dependencies are added:

```
//The plugin dependencies are added first to the classpath
    LinkedHashSet<File> resolvedJarArtifacts = new LinkedHashSet<>(getJars(configuration.getDependencies()));
```
Then we loose the this:

C:\Users\eraonel\git\tutorials\annotations\annotation-processing\target\classes

Anyone able to explain how this is added to classpath and how? Or do I need to change something in pom.xml?

**Reply by mpet** [reply]
My question is actually answered and I open another discussion about the issue mentioned above.
